✨ Support BYO Public IPv4 Pool when provision infrastructure (#4905)
kubernetes-sigs/cluster-api-provider-aws: docs/book/src/topics/bring-your-own-aws-infrastructure.md

Once the user has created externally managed AWSCluster, it is not allowed to convert it to CAPA managed cluster. However, converting from managed to externally managed is allowed.

User should only use this feature if their cluster infrastructure lifecycle management has constraints that the reference implementation does not support. See [user stories](https://github.com/kubernetes-sigs/cluster-api/blob/10d89ceca938e4d3d94a1d1c2b60515bcdf39829/docs/proposals/20210203-externally-managed-cluster-infrastructure.md#user-stories) for more details.

## Bring your own (BYO) Public IPv4 addresses

Cluster API also provides a mechanism to allocate Elastic IP from the existing Public IPv4 Pool that you brought to AWS[1].

> Bringing your own Public IPv4 Pool (BYOIPv4) can be used as an alternative to IPv4 charges introduced by AWS in February 2024[2].

Reviewer: ```suggestion
Bringing your own Public IPv4 Pool (BYOIPv4) can be used as an alternative to buying Public IPs from AWS, also considering the changes in charging for this since February 2024[2].
```

Supported resources to BYO Public IPv4 Pool (`BYO Public IPv4`):
- Nat Gateways
- Network Load Balancer for API server
- Machines

> Use `BYO Public IPv4` when you have brought to AWS custom IPv4 CIDR blocks and want to cluster automatically from custom pools instead of Amazon-provided pools.

Reviewer: ```suggestion
Use `BYO Public IPv4` when you have brought to AWS custom IPv4 CIDR blocks and want the cluster to automatically use IPs from the custom pool instead of Amazon-provided pools.
```

[1] https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-byoip.html
[2] https://aws.amazon.com/blogs/aws/new-aws-public-ipv4-address-charge-public-ip-insights/

### Prerequisites and limitations for BYO Public IPv4 Pool

- BYOIPv4 is limited to AWS to selected regions. See more in [AWS Documentation for Regional availability](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-byoip.html#byoip-reg-avail)
- The IPv4 address must be provisioned and advertised to the AWS account before the cluster is installed

> - The public IPv4 addresses is limited to the Network Border Group that the CIDR block have been provisioned, and it must be the same of the cluster will be installed.

Reviewer: Not sure what you mean here, we can probably rephrase

> - Only NAT Gateways and Netowork Load Balancer for API server will consume from the IPv4 pool defined in the network scope.

Reviewer: ```suggestion
- Only NAT Gateways and the Network Load Balancer for API server will consume from the IPv4 pool defined in the network scope.
```

Once the user has created externally managed AWSCluster, it is not allowed to convert it to CAPA managed cluster. However, converting from managed to externally managed is allowed.

User should only use this feature if their cluster infrastructure lifecycle management has constraints that the reference implementation does not support. See [user stories](https://github.com/kubernetes-sigs/cluster-api/blob/10d89ceca938e4d3d94a1d1c2b60515bcdf39829/docs/proposals/20210203-externally-managed-cluster-infrastructure.md#user-stories) for more details.

## Bring your own (BYO) Public IPv4 addresses

Cluster API also provides a mechanism to allocate Elastic IP from the existing Public IPv4 Pool that you brought to AWS[1].

Bringing your own Public IPv4 Pool (BYOIPv4) can be used as an alternative to buying Public IPs from AWS, also considering the changes in charging for this since February 2024[2].

Supported resources to BYO Public IPv4 Pool (`BYO Public IPv4`):

> - Nat Gateways

Reviewer 2: ```suggestion
- NAT Gateways
```
